dali/pipeline/operator/builtin/external_source.cu

```diff
@@ -52,18 +52,55 @@ void ExternalSource<GPUBackend>::RunImpl(DeviceWorkspace &ws) {
   {
     std::unique_lock<std::mutex> busy_lock(busy_m_);
     tensor_list_elm = tl_data_.PopFront();
-    internal_copy_to_storage = copy_to_storage_events_.PopFront();
-    cuda_event = cuda_events_.GetEmpty();
+    // even with no_copy we may have copied from TensorVector to TensorList and we
+    // need to sync with that
+    if (!no_copy_ || !tensor_list_elm.front()->shares_data()) {
+      internal_copy_to_storage = copy_to_storage_events_.PopFront();
+      if (!no_copy_) {
+        cuda_event = cuda_events_.GetEmpty();
+      }
+    }
   }
 
   auto &output = ws.Output<GPUBackend>(0);
-  cudaStream_t stream_used = ws.has_stream() ? ws.stream() : 0;
-  CUDA_CALL(cudaStreamWaitEvent(stream_used, *internal_copy_to_storage.front(), 0));
-  output.Copy(*(tensor_list_elm.front()), stream_used);
-  // record an event so Recycle can synchronize on it
-  cudaEventRecord(*cuda_event.front(), stream_used);
-  sync_worker_.DoWork(RecycleFunctor{this, std::move(cuda_event), std::move(tensor_list_elm),
-                                     std::move(internal_copy_to_storage)});
+  if (!no_copy_) {
+    cudaStream_t stream_used = ws.has_stream() ? ws.stream() : 0;
+    CUDA_CALL(cudaStreamWaitEvent(stream_used, *internal_copy_to_storage.front(), 0));
+    output.Copy(*(tensor_list_elm.front()), stream_used);
+    // record an event so Recycle can synchronize on it
+    cudaEventRecord(*cuda_event.front(), stream_used);
+    sync_worker_.DoWork(RecycleFunctor{this, std::move(cuda_event), std::move(tensor_list_elm),
+                                       std::move(internal_copy_to_storage)});
+  } else {
+    // if we copied from GPU TensorVector to TensorList we need to sync with that copy
+    if (!tensor_list_elm.front()->shares_data()) {
```

Comment:
That can probably be also indicated in the state?

Reply:
Done